### PR TITLE
fix(platform-browser): loosen KeyEventsPlugin.addEventListener element type

### PR DESCRIPTION
**Pull request by MatthiasKunnen**
## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?
`keyEventsPlugin.addEventListener` does not accept document as first parameter according to the type. This despite the fact that the function receives [`HTMLDocument`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDocument) in case of `(document:keydown.escape)="..."`

## What is the new behavior?
Loosened the type of the parameter to accept [`Node`](https://developer.mozilla.org/en-US/docs/Web/API/Node), a type both [`HTMLElement`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement) and [`HTMLDocument`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDocument) have in common. 

Alternatively, we could go with `element: HTMLElement | HTMLDocument`.

The new type `Node` is based on the most specific type I could find traversing the function, `BrowserDomAdapter.onAndCancel()`.

https://github.com/angular/angular/blob/f8096d499324cf0961f092944bbaedd05364eea1/packages/platform-browser/src/browser/browser_adapter.ts#L124

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by MatthiasKunnen** [comment]
I've updated the PR to resolve merge conflicts.

**Review by pkozlowski-opensource — CHANGES_REQUESTED** [review]
@MatthiasKunnen hey, thnx for the PR - I do agree that `HTMLElement` is not the most appropriate type but 
 changing it to `Node` is not correct since we can't add event listeners to Nodes (https://developer.mozilla.org/en-US/docs/Web/API/Node).

The appropriate type here would be [EventTarget](https://github.com/microsoft/TypeScript/blob/76115794215135716f10d8e8b98e82be616e0d6d/lib/lib.dom.d.ts#L5398-L5422) I guess. Having said this we would have to adjust types in other places (`EventManagerPlugin` and all its implementations) to be consistent.

Long story short:
* the prosed change is not correct (one can't add listeners on `Node`  and the fact that code after the proposed change compiles is just... sloppy typings in other places....
* we would need use the  `EventTarget` and change all relevant places.

Please let me know if you want to continue work towards making those types correct (I think that we can make it a backward compatible change as every `HTMLElement` is `EventTarget`

**Comment by alfaproject** [comment]
@pkozlowski-opensource `Node` extends from `EventTarget` tho

**Comment by pkozlowski-opensource** [comment]
> Node extends from EventTarget tho

@alfaproject oh, indeed!
* https://developer.mozilla.org/en-US/docs/Web/API/Node
* https://github.com/microsoft/TypeScript/blob/76115794215135716f10d8e8b98e82be616e0d6d/lib/lib.dom.d.ts#L10740

Good catch @alfaproject ! Still, my comments are still valid, but you are right!

**Comment by MatthiasKunnen** [comment]
This is the inheritance diagram of both `HTMLElement` and `HTMLDocument`.

image

Based on that I believe we can either use the closest common ancestor, `Node`, or `HTMLDocument | HTMLElement`. The latter will probably be better for consumers of the method.
I could go in and see what needs updating if we can decided upon the aforementioned type.

**Comment by pkozlowski-opensource** [comment]
> Based on that I believe we can either use the closest common ancestor, Node, or HTMLDocument | HTMLElement.

@MatthiasKunnen couldn't we use `EventTarget` which is the most "specific" type we need here?

**Comment by MatthiasKunnen** [comment]
I might misunderstand you but this is how I see it. The _most_ specific type is `HTMLDocument | HTMLElement`, the most specific _common_ type is `Node`, the _least_ specific _common_ type is `EventTarget`.

In my opinion, choosing the most specific type is best since it will require the least type checking by users of the method.
